### PR TITLE
add alignment tools

### PR DESCRIPTION
**Pull request by grimoire**
This tool can be used to align the rewrite module to the origin module.
Provide layout conversion, input/output extractor and context maker.

### CODE REVIEW

**Comment by lvhan028** [comment]
Please briefly describe the motivation and the usage of the tools

File: docs/en/advance/pytorch_new_model.md [hunk 1]
```diff
@@ -243,6 +243,72 @@ MODEL_MAP.update({
...
+input_ids = torch.tensor([[1, 2, 3, 4, 5]]).cuda()
+extractor = ModuleIOExtractor(model, self_attn)
+attn_args, attn_kwargs, attn_output = extractor.extract(input_ids)
+```
+
+The inputs of rewrite module are different from the inputs of origin module:
+
+1. Module requires some special inputs, which are passed through `StepContext`. We can create one with `make_step_context`.
+2. `input_ids`, `hidden_states` should be continuous. We can use `continuous_tensor` to do the process.
+3. `past_key_value` should be paged to meet the demond of paged attention.
```
Comment (RunningLeon) [inline docs/en/advance/pytorch_new_model.md:278]: demand

File: docs/en/advance/pytorch_new_model.md [hunk 2]
```diff
@@ -243,6 +243,72 @@ MODEL_MAP.update({
...
+
+```python
+from lmdeploy.pytorch.models import patch
+
+# patch and test
+patched_self_attn = patch(self_attn, extra_args=['context'])
+patched_output = patched_self_attn.patched_forward(*attn_args,
+                                                    **attn_kwargs,
+                                                    context=context)
+torch.testing.assert_close(patched_output[0],
```
Comment (RunningLeon) [inline docs/en/advance/pytorch_new_model.md:306]: here fails. Is it meant to ?
```
Traceback (most recent call last):
  File "t.py", line 39, in <module>
    torch.testing.assert_close(patched_output[0],
  File "/opt/py38/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1520, in assert_close
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 13 / 20480 (0.1%)
Greatest absolute difference: 2.6702880859375e-05 at index (0, 4, 3443) (up to 1e-05 allowed)
Greatest relative difference: 0.07958984375 at index (0, 2, 124) (up to 0.001 allowed)
```

**Review by RunningLeon — APPROVED** [review]
LGTM

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**